database/tracer.go

```diff
@@ -99,4 +106,111 @@ func (t ethDBTracer) AssertPersistedGenesisBlock(genesis core.Genesis) {
 			"persisted_balance", persistedBalance.String(),
 		)
 	}
+}
+
+func (t EthDBTracer) AssertPostTxSimulationState(from common.Address, tx *types.Transaction) {
+	t.logger.Infow("Tracing if ETH DB is in a valid state after simulation...", "chaindata", t.chainDataDir)
+
+	chainDb, err := ethdb.NewLDBDatabase(t.chainDataDir, 0, 0)
+	if err != nil {
+		t.logger.Errorw("unable to open LDB db", "err", err)
+		return
+	}
+	defer chainDb.Close()
+
+	headHash := rawdb.ReadHeadBlockHash(chainDb)
+	headNumber := rawdb.ReadHeaderNumber(chainDb, headHash)
+	block := rawdb.ReadBlock(chainDb, headHash, *headNumber)
+
+	stateDB, err := state.New(block.Root(), state.NewDatabase(chainDb))
+	if err != nil {
+		t.logger.Errorw("unable to open new stateDB", "err", err)
+		return
+	}
+
+	// Given the same TX inputs, hash must not change
+	expectedTxHash := common.HexToHash("0x49828264ee35226d1242893343c29537a04f61bbb2671d58d7ff9e16f5e5c4bd")
+	if !bytes.Equal(block.TxHash().Bytes(), expectedTxHash.Bytes()) {
+		t.logger.Errorw("incorrect simulated TX hash", "expected", expectedTxHash.Hex(), "actual", block.TxHash().Hex())
+	} else {
+		t.logger.Infow("correct simulated TX hash", "hash", block.TxHash().Hex())
+	}
+
+	// Given the same TX changes, block root hash must not change
+	expectedRootHash := common.HexToHash("0x39e5006078070400d77b2494c879d5a6315ab3a7ae19c1309c0b4126f1f56f9c")
+	if !bytes.Equal(block.Root().Bytes(), expectedRootHash.Bytes()) {
+		t.logger.Errorw("incorrect root hash", "expected", expectedRootHash.Hex(), "actual", block.Root().Hex())
+	} else {
+		t.logger.Infow("correct root hash", "hash", block.Root().Hex())
+	}
+
+	// Coinbase should be 0x0
+	expectedCoinbase := common.HexToAddress("0x0")
+	if !bytes.Equal(block.Coinbase().Bytes(), expectedCoinbase.Bytes()) {
+		t.logger.Errorw("incorrect coinbase", "expected", expectedCoinbase.Hex(), "actual", block.Coinbase().Hex())
+	} else {
+		t.logger.Infow("correct coinbase", "acc", block.Coinbase().Hex())
+	}
+
+	// Parent hash is always the genesis block
+	expectedParentHash := common.HexToHash("0x55e06fc7b51b31efb053f128068be8b09c86569895d98591ea5790b683770c58")
+	if !bytes.Equal(block.ParentHash().Bytes(), expectedParentHash.Bytes()) {
+		t.logger.Errorw("incorrect parent hash", "expected", expectedParentHash.Hex(), "actual", block.ParentHash().Hex())
+	} else {
+		t.logger.Infow("correct parent hash", "hash", block.ParentHash().Hex())
+	}
+
+	genesisFromBalance, _ := new(big.Int).SetString("300000000000000000000000000", 10)
```

Comment:
[minor] Probably you could have used the `WeiToPht` method, currently on `Prometheus/utils` to simplify and unify token<>wei conversions

Reply:
ok